src/adapter/src/subscribe.rs

```diff
@@ -94,8 +105,105 @@ impl ActiveSubscribe {
+                                    })
+                                });
+                            }
+                            SubscribeOutput::EnvelopeUpsert { key_indices } => {
+                                let mut left_datum_vec = mz_repr::DatumVec::new();
+                                let mut right_datum_vec = mz_repr::DatumVec::new();
+                                let order_by = &key_indices
+                                    .iter()
+                                    .map(|idx| ColumnOrder {
+                                        column: *idx,
+                                        desc: false,
+                                        nulls_last: true,
+                                    })
+                                    .collect_vec();
+                                rows.sort_by(|(left_time, left_row, left_diff), (right_time, right_row, right_diff)| {
+                                    left_time.cmp(right_time).then_with(|| {
+                                        let left_datums = left_datum_vec.borrow_with(left_row);
+                                        let right_datums = right_datum_vec.borrow_with(right_row);
+                                        compare_columns(order_by, &left_datums, &right_datums, || {
+                                            left_row.cmp(right_row).then(left_diff.cmp(right_diff))
+                                        })
+                                    })
+                                });
+
+                                let mut new_rows = Vec::new();
+                                let mut it = rows.iter();
+                                let mut datum_vec = mz_repr::DatumVec::new();
+                                while let Some(start) = it.next() {
+                                    let group =
+                                        std::iter::once(start).chain(it.take_while_ref(|row| {
+                                            let left_datums = left_datum_vec.borrow_with(&start.1);
+                                            let right_datums = right_datum_vec.borrow_with(&row.1);
+                                            start.0 == row.0
+                                                && compare_columns(
+                                                    order_by,
+                                                    &left_datums,
+                                                    &right_datums,
+                                                    || Ordering::Equal,
+                                                ) == Ordering::Equal
+                                        }));
+                                    let mut saw_new_row = false;
+                                    for (t, r, d) in group {
+                                        // We don't have to consider (k, v, -1), (k, v, +1) because
+                                        // the rows are already consolidated. For a key, if all we
```

Comment:
I don't see where consolidation happened. Is it an effect of the group thing above?

Reply:
That's guaranteed by https://github.com/MaterializeInc/materialize/blob/3e77515eba1c96582deeb1a0267d532f6d29dacc/src/compute-client/src/protocol/response.rs#L107 which i added in a recent PR (https://github.com/MaterializeInc/materialize/commit/2072183bb3d6fb8d03ba07795d8755e6e4ef2bfb)